File: frontend/src/lib/types/api-keys.ts
import type { UUID } from "@/lib/types/common";

/**
 * API key wire types, hand-mirrored from `app/schemas/api_keys.py`. Keys are
 * the credential an agent harness holds for a collection's MCP endpoint.
 */

/** Mirrors `app/schemas/enums.py::ApiKeyCapability`. */
export type ApiKeyCapability = "tools:invoke" | "files:read" | "files:write";

/** Mirrors `ApiKeyRead` — never carries the secret. */
export interface ApiKey {
  id: UUID;
  name: string;
  prefix: string;
  capabilities: ApiKeyCapability[];
  all_collections: boolean;
  collection_ids: UUID[];
  created_at: string;
  last_used_at: string | null;
  expires_at: string | null;
  revoked_at: string | null;
}

/** Mirrors `ApiKeyCreate`. */
export interface ApiKeyCreatePayload {
  name: string;
  capabilities: ApiKeyCapability[];
  all_collections?: boolean;
  collection_ids?: UUID[];
  expires_in_days?: number | null;
}

/** Mirrors `ApiKeyCreated` — the one shape carrying the plaintext secret. */
export interface ApiKeyCreated {
  key: ApiKey;
  secret: string;
}

/** Mirrors `ApiKeyList`. */
export interface ApiKeyListResponse {
  keys: ApiKey[];
}

File: app/schemas/api_keys.py
"""API key wire types.

The secret appears in exactly one response shape (`ApiKeyCreated`) and never
again: `ApiKeyRead` carries only the non-secret display prefix, so a key
listing can never leak a usable credential.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.enums import ApiKeyCapability

#: Capability sets are small by design; the cap stops a pathological payload.
MAX_CAPABILITIES = len(ApiKeyCapability)


class ApiKeyRead(BaseModel):
    """One API key as the management UI sees it — never the secret."""

    id: UUID
    name: str
    prefix: str
    capabilities: list[ApiKeyCapability]
    collection_ids: list[UUID]
    created_at: datetime
    last_used_at: datetime | None = None
    expires_at: datetime | None = None
    revoked_at: datetime | None = None


class ApiKeyCreate(BaseModel):
    """Payload for issuing a key: what it may do, and which collections it reaches.

    `collection_ids` is always explicit — there is no "every collection" grant,
    so a key's reach cannot silently grow to cover collections created after it.
    """

    name: str = Field(min_length=1, max_length=100)
    capabilities: list[ApiKeyCapability] = Field(min_length=1, max_length=MAX_CAPABILITIES)
    collection_ids: list[UUID] = Field(min_length=1, max_length=200)
    expires_in_days: int | None = Field(default=None, ge=1, le=3650)

    @field_validator("capabilities")
    @classmethod
    def _unique_capabilities(cls, value: list[ApiKeyCapability]) -> list[ApiKeyCapability]:
        """Reject duplicate capabilities rather than silently collapsing them."""
        if len(set(value)) != len(value):
            raise ValueError("Capabilities must be unique.")
        return value


class ApiKeyCreated(BaseModel):
    """The one response that carries the plaintext secret."""

    key: ApiKeyRead
    secret: str


class ApiKeyList(BaseModel):
    """A user's API keys."""

    keys: list[ApiKeyRead]
